Features/#377 grid separation

khelfen opened the pull request:
# Description

- [x] add function to separate lv grids
- [x] add methodology to decide which lv grids to separate

Fixes #377 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] New and adjusted code is formatted using the `pre-commit` hooks
- [x] New and adjusted code includes type hinting now
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] The Read the Docs documentation is compiling correctly
- [x] If new packages are needed, I added them the [setup.py](https://github.com/openego/eDisGo/blob/dev/setup.py), and if needed the [rtd_requirements.txt](https://github.com/openego/eDisGo/blob/dev/rtd_requirements.txt), the [eDisGo_env.yml](https://github.com/openego/eDisGo/blob/dev/eDisGo_env.yml) and the [eDisGo_env_dev.yml](https://github.com/openego/eDisGo/blob/dev/eDisGo_env_dev.yml).
- [x] I have added new features to the corresponding [whatsnew](https://github.com/openego/eDisGo/tree/dev/doc/whatsnew) file

Commit:
added lv grid separation functionality

Commit:
added tests for lv grid separation

Commit:
Merge branch 'dev' into features/#377-grid-separation

Commit:
Merge branch 'dev' of github.com:openego/eDisGo into features/#377-grid-separation

birgits commented on edisgo/flex_opt/reinforce_measures.py:
```diff
@@ -725,3 +732,379 @@ def _replace_by_parallel_standard_lines(lines):
+        logger.info(f"{grid}==>method:add_station_at_half_length is running ")
+        # Create the bus-bar name of primary and secondary side of new MV/LV station
+        lv_bus_new = create_bus_name(station_node, "lv")
+        mv_bus_new = create_bus_name(station_node, "mv")
```
New transformer should be connected to existing MV bus and no new MV created.

khelfen replied:
Done in [e720b79](https://github.com/openego/eDisGo/pull/380/commits/e720b79abfa424fcccfd54915c8a9bca4f6809e0)

birgits commented on edisgo/flex_opt/reinforce_measures.py:
```diff
@@ -725,3 +732,379 @@ def _replace_by_parallel_standard_lines(lines):
+        }
+    """
+
+    def get_weight(u, v, data: dict) -> float:
```
Parameters u and v are unused. Are they needed?

khelfen replied:
The `get_weight` function is given to the `networkx.algorithms.shortest_paths.weighted._dijkstra` function. This was copied from already existing parts in the code. E.g.:

https://github.com/openego/eDisGo/blob/c6a2edb6db4c55ff81b6ebe9bb216035665d1da3/edisgo/flex_opt/reinforce_measures.py#L412

I didn't check what the dijkstra does with the given function, but I think a simplification creates more overhead than it's worth.

khelfen replied:
The `get_weight` function is given to the `networkx.algorithms.shortest_paths.weighted._dijkstra` function. This was copied from already existing parts in the code. E.g.:

https://github.com/openego/eDisGo/blob/c6a2edb6db4c55ff81b6ebe9bb216035665d1da3/edisgo/flex_opt/reinforce_measures.py#L412

I didn't check what the dijkstra does with the given function, but I think a simplification creates more overhead than it's worth.

birgits commented on edisgo/flex_opt/reinforce_measures.py:
```diff
@@ -725,3 +732,379 @@ def _replace_by_parallel_standard_lines(lines):
+            bus = bus.split("_")
+            grid_id_ind = bus.index(str(grid.id))
+            # TODO: how to name new grids?
+            bus[grid_id_ind] = f"{grid.id}1001"
```
I think the naming is okay. The ID currently needs to be something that can be converted to an integer. Adding a check for uniquness would be good.

khelfen replied:
I changed this with the recent update, because in the case of multiple separation of the same lv grid this created duplicates. New code:

https://github.com/openego/eDisGo/blob/a6793c222ee7d9ec1387365a6927997a4709024e/edisgo/flex_opt/reinforce_measures.py#L1030-L1043

birgits commented on edisgo/flex_opt/reinforce_measures.py:
```diff
@@ -725,3 +732,379 @@ def _replace_by_parallel_standard_lines(lines):
+        first_line = relevant_lines[relevant_lines.bus1 == first_node].index[0]
+
+        # the last node of the feeder
+        last_node = nodes_feeder[-1]
```
I am not sure if that is correct. nodes_feeder contains the buses in the feeder in random order, doesn't it?

khelfen replied:
As `nx.shortest_path` 

https://github.com/openego/eDisGo/blob/ac17817920b458a6bcef201dc54f3afee30c0762/edisgo/flex_opt/reinforce_measures.py#L945

returns the nodes in order from first to last I replaced the nodes in the `first_nodes_feeders` values with the sorted nodes of the main feeder determined by number of nodes in `paths`:

https://github.com/openego/eDisGo/blob/ac17817920b458a6bcef201dc54f3afee30c0762/edisgo/flex_opt/reinforce_measures.py#L969-L976

This way `node_1_2` is guaranteed to be part of the main path as well. In combination with `get_downstream_buses`

https://github.com/openego/eDisGo/blob/ac17817920b458a6bcef201dc54f3afee30c0762/edisgo/flex_opt/reinforce_measures.py#L1032

it is still ensured that all nodes of branched of parts of the feeder are relocated to the new grid.

birgits commented on edisgo/flex_opt/reinforce_measures.py:
```diff
@@ -725,3 +732,379 @@ def _replace_by_parallel_standard_lines(lines):
+
+    for first_node, nodes_feeder in first_nodes_feeders.items():
+        # first line of the feeder
+        first_line = relevant_lines[relevant_lines.bus1 == first_node].index[0]
```
The first node could also be in bus0!

khelfen replied:
Fixed:

https://github.com/openego/eDisGo/blob/e720b79abfa424fcccfd54915c8a9bca4f6809e0/edisgo/flex_opt/reinforce_measures.py#L975-L977

birgits commented on edisgo/flex_opt/reinforce_measures.py:
```diff
@@ -725,3 +732,379 @@ def _replace_by_parallel_standard_lines(lines):
+        #  connected to the station) feeder cannot be split. Instead, every second
+        #  inept feeder is assigned to the new grid
+        if node_1_2 not in first_nodes_feeders or count_inept % 2 == 1:
+            nodes_tb_relocated[node_1_2] = nodes_feeder[nodes_feeder.index(node_1_2) :]
```
This will only be correct, if nodes_feeder is sorted by distance to station, which I am not sure it is..

khelfen replied:
Changed this to make sure to only use downstream buses:

https://github.com/openego/eDisGo/blob/80fcbaea110ea82daecf9e076b5cea900a0c2d95/edisgo/flex_opt/reinforce_measures.py#L1022

birgits commented on edisgo/flex_opt/reinforce_measures.py:
```diff
@@ -725,3 +732,379 @@ def _replace_by_parallel_standard_lines(lines):
+        transformers_changes.update(transformer_changes)
+
+        logger.debug(
+            f"{edisgo_obj.topology.mv_grid}==>A new grid {lv_grid_id_new} "
```
Grid ID is automatically added to logging, so this can be removed.

khelfen replied:
Changed it

Commit:
Added workflow to separate all lv grids

Workflow separates lv grids if a specific overloading threshold is reached. LV grids
can be separated multiple times if necessary. Workflow is now a selectable part of
the enhanced reinforcement method. Added simple tests.

Commit:
connect new lv grid to existing mv bus instead of creating a new mv bus

Commit:
make sure that only downstream buses of bus_1_2 are relocated to th new lv grid

Commit:
change logging string

Commit:
Make sure that the nodes of a feeder are sorted correctly when separating a grid

Also make sure to only use nodes from the main feeder of a feeder.

Commit:
add equipment changes to workflow

Commit:
adapted whatsnew

birgits commented on edisgo/flex_opt/reinforce_grid.py:
```diff
@@ -861,30 +853,40 @@ def enhanced_reinforce_grid(
         edisgo_obj.reinforce(mode=None, catch_convergence_problems=True, **kwargs)
         logger.info("Initial enhanced reinforcement succeeded.")
-    except:  # noqa: E722
+    except ValueError:
```
Non-convergence issues can also lead to a RuntimeError, wherefore just catching ValueError is not sufficient.

khelfen replied:
`RuntimeError` s are now excepted as well:

https://github.com/openego/eDisGo/blob/92c82a520d0c1f3249fdcbd4ba0c8b77f09c66d2/edisgo/flex_opt/reinforce_grid.py#L881

Commit:
docstrings

Commit:
also except RuntimeErrors in enhanced reinforcement

Commit:
fix docstrings

Commit:
Merge branch 'dev' into features/#377-grid-separation

birgits commented on edisgo/flex_opt/reinforce_grid.py:
```diff
@@ -967,3 +979,168 @@ def enhanced_reinforce_grid(
+                    .astype(float)
+                )
+
+                if tech == "storage_units":
```
I am not quite sure if storage units are handled correctly like this.. Wouldn't it be easier to take np.hypot after adding up all active and reactive powers of loads, generators and storage units?

khelfen replied:
Yeah, you are right. Somehow my thinking was far too complicated. I changed it now:

https://github.com/openego/eDisGo/blob/f473b5ac6990fe55a5b46f1939769ff296d2b6c3/edisgo/flex_opt/reinforce_grid.py#L1031-L1069

birgits commented on edisgo/flex_opt/reinforce_measures.py:
```diff
@@ -725,3 +733,434 @@ def _replace_by_parallel_standard_lines(lines):
+    G = grid.graph
+
+    # main station
+    station_node = list(G.nodes)[0]
```
Does the station bus always have to be the first node? I am not sure, as G is an undirected and unordered graph.

khelfen replied:
`station_node` is no identified from the `transformers_df`:

https://github.com/openego/eDisGo/blob/48255d3f4ed21f750fb4b0ddb76791d1bd28c660/edisgo/flex_opt/reinforce_measures.py#L925

birgits commented on edisgo/flex_opt/reinforce_measures.py:
```diff
@@ -725,3 +733,434 @@ def _replace_by_parallel_standard_lines(lines):
+
+        path = nx.shortest_path(G, station_node, node)
+
+        for first_node in relevant_lines.bus1.values:
```
I don't really understand what happens in this and the outer for loop, so a comment would be nice. And I was also wondering if it is necessary that bus1 is always the downstream node? Because here it could also be the station node and I don't know if that would be okay or not.

khelfen replied:
I changed `relevant_lines.bus1.values` to

https://github.com/openego/eDisGo/blob/3a28aa280d8569745d4ded2c11f5ccb906420da3/edisgo/flex_opt/reinforce_measures.py#L931-L933

and added some comments.

Commit:
Minor doc and logging changes

Commit:
simplify apparent power calculation

Commit:
make sure to idntify station node correctly

Commit:
add comments at hard to undrstand parts

Commit:
fix test_separate_lv_grid

Commit:
Minor doc change

Commit:
Bug fix avoid float becoming object type

Commit:
Correct docstring

Commit:
Fix check if LV grid ID is provided

Commit:
Minor logging typo fix

Commit:
Add grid separation when power flow for LV grid does not converge
